### DISCUSSION TITLE
Control Heating on Thermostat Junkers FW120, Change Program, Call next Program

### DISCUSSION THREAD

**Discussion by bergteam** [emsesp/EMS-ESP32#1687] [General]
I am using a Junkers Boiler with FW120 Thermostat. The goal is to temporarily control heating through iobroker.

**What is working so far:**
+ set "hc1 mode (mode)" to nofrost/eco/heat to control and once finished return control back to FW120 with mode=auto
+ change "hc1 nofrost/eco/heat temperature", this sets the temperature permanently

**1. Was the following already considered for implementation?**
+ on the FW120 there is a button to "temporarily advance to next timewindow", see the attached screenshot from the manual
+ if not, may I trace the messages and post here, during pressing of the button?

![FW120-CallNextTimeWindow](https://github.com/emsesp/EMS-ESP32/assets/153366337/ecb7fd47-0b38-4d35-b6b7-25e896b7083c)

![FW120-CallNextTimeWindow2](https://github.com/emsesp/EMS-ESP32/assets/153366337/4c3084cf-6b3e-40c3-9398-188d7d4286bc)

**2. What can be done with the setting "hc1 program (program)"?**
+ it is alwas set to "prog a", regardless which mode and timewindow (heat/nofrost) the FW automatically selects
+ can this be used to set the FW120 program or timewindow from outside?

**Comment by MichaelDvP** [comment]
If i remember correctly the temporary selected temperature is only internal of the thermostat, and is not published to the bus (only the monitor shows new setpoint. But maybe this is changed with Junkers firmware, try to log a change. 
Also log and push the next button, i think we have never tested this before.

Some Junkers thermostats have different porgrams a-f, and you can choose one of them to be active. The programs are stored in telegram 0x22C (a), 0x22D (b), etc. for hc1 and 0x233 (a), 0x234 (b) for hc2.  Just try to change to a different program and see if the themostat accept it. Or check in terminal with `read 10 22`D,  etc. if the telegrams exist.

**Reply by bergteam** [reply]
Find attached the #watch 0x10 log on the FW120 thermostat, while pressing the button.

PressButton, Changes ModeType=>NoFrost, Heating was already off due to outside Temp > HeatingOffTemp
010+23:08:02.768 N 45: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01 02 00 32 01 16 F3 34 00
010+23:08:03.271 N 46: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 01 (offset 4)
010+23:08:03.300 N 47: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01

PressButton, Changes ModeType=>NoFrost, Heating was already off due to outside Temp > HeatingOffTemp
003+05:58:43.414 N 35: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01 02 00 32 01 04 F3 34 05
003+05:58:43.917 N 36: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 01 (offset 4)
003+05:58:43.945 N 37: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01

PressButton, Changes ModeType=>NoFrost, Heating was on and immediately switched off
003+21:24:22.937 N 7: [emsesp] thermostat(0x10) -B-> All(0x00), RCTime(0x06), data: 18 04 06 0A 2A 0E 03 01
003+21:24:24.295 N 8: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01 02 00 32 01 22 F3 34 00
003+21:24:24.748 N 9: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 01 (offset 4)
003+21:24:24.771 N 10: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 00
003+21:24:24.797 N 11: [emsesp] thermostat(0x10) -W-> boiler(0x08), JunkersSetMixer(0x23), data: 00
003+21:24:24.826 N 12: [emsesp] thermostat(0x10) -B-> All(0x00), JunkersMonitor(0x016F), data: 01
003+21:24:25.251 N 13: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBAFlags(0x35), data: 11 00 26 (offset 1)
003+21:24:25.296 N 14: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 00
003+21:24:25.321 N 15: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 00 (offset 2)
003+21:24:25.350 N 16: [emsesp] thermostat(0x10) -W-> boiler(0x08), JunkersSetMixer(0x23), data: 00 00 00
003+21:24:25.371 N 17: [emsesp] thermostat(0x10) -W-> boiler(0x08), UBASetPoints(0x1A), data: 01 (offset 4)

**Reply by MichaelDvP** [reply]
Seems there is nothing stored in thermostat, it's just commands to the boiler. I guess:
0x1A offset 4 sets the nofrost mode. 
0x1A is a command telegram and can not be read, we dont know how many bytes it can hold and new only the first 3, which are set in the last section:
`UBASetPoints(0x1A), data: 00` boiler set-temperature 0
`UBASetPoints(0x1A), data: 00 (offset 2)` burner power for dhw 0
There is also a command for DHW:
`UBAFlags(0x35), data: 11 00 26 (offset 1)` dhw and circulation allowed, no offset, setpoint 38°C
 
If you press the button in nofrost mode to switch to heat (or whatt is next mode), what is the command? I guess it's also 0x1A offset 4 with a different value.

**Comment by bergteam** [comment]
Regarding the scheduler on the FW120.

1. Mode can be selected individually on the following day groups
AllDays, Mo-Fr, Sa-So, Mo, Tu, We, Th, Fr, Sa, Su

2. On each Group 6 timing blocks (assign mode to time) exist => correlate to "prog a..f" shown in ems-esp
![FW120_programs](https://github.com/emsesp/EMS-ESP32/assets/153366337/ee270150-cfa2-4019-ae77-8a04b0197393)

Find below the reading of my FW120. I can not yet decode the result myself.
Settings:
```
P1 Heating 04:00
P2 NoFrost 10:00
P3 Heating 15:00
P4 NoFrost 18:00

```
Readout:
```
ems-esp:$ read 10 22C
003+22:08:47.679 N 48: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022C), data: <empty>
ems-esp:$ read 10 22D
003+22:08:55.357 N 49: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022D), data: 02 10 00 28 02 3C 00 48 FF FF FF FF 02 10 00 28 02 3C 00 48 FF FF FF FF 02
003+22:08:55.456 N 50: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022D), data: 10 00 28 02 3C 00 48 FF FF FF FF 02 10 00 28 02 3C 00 48 FF FF FF FF 02 10 (offset 25)
003+22:08:55.554 N 51: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022D), data: 00 28 02 3C 00 48 FF FF FF FF 02 10 00 28 02 3C 00 48 FF FF FF FF 02 10 00 (offset 50)
003+22:08:55.675 N 52: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022D), data: 28 02 3C 00 48 FF FF FF FF 50 72 6F 67 72 61 6D 6D 20 41 20 20 20 20 20 20 (offset 75)
003+22:08:55.757 N 53: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022D), data: 20 20 (offset 100)
ems-esp:$ read 10 22E
003+22:08:58.707 N 54: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022E), data: 02 18 01 20 02 44 00 58 FF FF FF FF 02 18 01 20 02 44 00 58 FF FF FF FF 02
003+22:08:58.805 N 55: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022E), data: 18 01 20 02 44 00 58 FF FF FF FF 02 18 01 20 02 44 00 58 FF FF FF FF 02 18 (offset 25)
003+22:08:58.903 N 56: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022E), data: 01 20 02 44 00 5E FF FF FF FF 02 1C 00 5E FF FF FF FF FF FF FF FF 02 20 00 (offset 50)
003+22:08:59.024 N 57: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022E), data: 58 FF FF FF FF FF FF FF FF 50 72 6F 67 72 61 6D 6D 20 42 20 20 20 20 20 20 (offset 75)
003+22:08:59.098 N 58: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022E), data: 20 20 (offset 100)
ems-esp:$ read 10 22F
003+22:09:02.580 N 59: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022F), data: 02 1C 01 5C FF FF FF FF FF FF FF FF 02 1C 01 5C FF FF FF FF FF FF FF FF 02
003+22:09:02.678 N 60: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022F), data: 1C 01 5C FF FF FF FF FF FF FF FF 02 1C 01 5C FF FF FF FF FF FF FF FF 02 1C (offset 25)
003+22:09:02.777 N 61: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022F), data: 01 5C FF FF FF FF FF FF FF FF 02 1C 01 5C FF FF FF FF FF FF FF FF 02 1C 01 (offset 50)
003+22:09:02.898 N 62: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022F), data: 5C FF FF FF FF FF FF FF FF 50 72 6F 67 72 61 6D 6D 20 43 20 20 20 20 20 20 (offset 75)
003+22:09:02.979 N 63: [emsesp] thermostat(0x10) -W-> Me(0x0B), ?(0x022F), data: 20 20 (offset 100)
ems-esp:$ read 10 22G
ems-esp:$ read 10 22H
```

**Reply by MichaelDvP** [reply]
Ok, we have "Programm A" in 0x022D, with:
6 setpoints per day, 2 bytes for each setpoint, first mode 02=heat, 00=nofrost, second time in steps of 15 min, 10=>16x15min=>04:00, 28->40x15min->10:00, etc., unset setpoins read FF FF
Mo: bytes 0-11, Tu: 12-23, etc: The groups in thermostat UI just set values for each day.
The last bytes are ASCII for the program name: 50 72 6F 67 72 61 6D 6D 20 41 20 20 20 20 20 20 20 20 => `Programm A        `

Since long time we are thinking how to display the programs, but 42 setpoints with each two entities and 3 programs gives 246 entities, blowing the memory of esp32. We've discussed to store only the telegrams and decode in a special program-editor webpags, but Bosch reinvents the program storage for each of their thermostats new, we have to build different decoders/settings for RC35, RC300, FW120, and some more.
